merge/merge.go

```diff
@@ -0,0 +1,171 @@
+package diff
+
+import (
+	"log"
+	"strconv"
+)
+
+const indicesReorderedKey = "$"
+
+// Merge merges an update diff into the previous version of a JSON object,
+// and returns the updated version.
+//
+// It effectively reverses the diff algorithm implemented in package diff.
+func Merge(prev map[string]interface{}, diff map[string]interface{}) interface{} {
+	new := make(map[string]interface{})
+
+	// Update exisiting fields.
+	for k, v := range prev {
+		d, ok := diff[k]
+		// No change.
+		if !ok {
+			new[k] = v
+			continue
+		}
+		// Removed.
+		if isRemoved(d) {
+			continue
+		}
+		// Updated.
+		var newV interface{}
+		switch v := v.(type) {
+		case map[string]interface{}:
+			newV = MergeMap(v, d)
+		case []interface{}:
+			newV = MergeArray(v, d)
+		default:
+			newV = MergeReplaced(d)
+		}
+		new[k] = newV
+	}
+
+	// Merge in new fields.
+	for k, d := range diff {
+		if _, ok := prev[k]; !ok {
+			new[k] = MergeReplaced(d)
+		}
+	}
+
+	return new
+}
+
+// MergeMap applies a delta to a map field.
+func MergeMap(prev map[string]interface{}, delta interface{}) interface{} {
```

Comment:
They don't need to be public anymore (after the restructuring); updated to be private.